pkg/kv/kvserver/rangefeed/unbuffered_registration.go

```diff
@@ -0,0 +1,383 @@
+// events in catch-up buffer while publishing them. Note that
+// discardCatchUpBuffer should never be called concurrently with this function.
+// Caller is responsible for draining it again if error is returned.
+func (ubr *unbufferedRegistration) publishCatchUpBuffer(ctx context.Context) error {
+	publish := func() error {
+		for {
+			select {
+			case e := <-ubr.mu.catchUpBuf:
+				err := ubr.stream.SendBuffered(e.event, e.alloc)
+				e.alloc.Release(ctx)
+				putPooledSharedEvent(e)
+				if err != nil {
+					return err
+				}
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				// Done.
+				return nil
+			}
+		}
+	}
+
+	if err := publish(); err != nil {
+		return err
+	}
+
+	ubr.mu.Lock()
+	defer ubr.mu.Unlock()
+
+	if err := publish(); err != nil {
+		return err
+	}
+
+	// Even if the catch-up buffer has overflowed, all events in it should still
+	// be published. Important to disconnect before setting catchUpBuf to nil.
+	if ubr.mu.catchUpOverflowed {
+		return newRetryErrBufferCapacityExceeded()
+	}
+
+	// Success.
+	ubr.mu.catchUpBuf = nil
+	ubr.mu.caughtUp = true
+	return nil
+}
+
+// discardCatchUpBuffer drains catchUpBuf and release all memory held by
+// the events in catch-up buffer without publishing them. It is safe to assume
+// that catch-up buffer is empty and nil after this.
+func (ubr *unbufferedRegistration) discardCatchUpBuffer(ctx context.Context) {
+	ubr.mu.Lock()
+	defer ubr.mu.Unlock()
+	// TODO(wenyihu6): Check if we can just discard without holding the lock first
+	// ? We shouldn't be reading from the buffer at the same time
+	ubr.discardCatchUpBufferWithLock(ctx)
+}
+
+func (ubr *unbufferedRegistration) discardCatchUpBufferWithLock(ctx context.Context) {
+	if ubr.mu.catchUpBuf == nil {
+		// Already drained.
+		return
+	}
+	func() {
+		for {
+			select {
+			case e := <-ubr.mu.catchUpBuf:
+				e.alloc.Release(ctx)
+				putPooledSharedEvent(e)
+			default:
+				// Done.
+				return
+			}
+		}
+	}()
+
+	ubr.mu.catchUpBuf = nil
+	ubr.mu.caughtUp = true
+}
+
+// detachCatchUpIter detaches the catchUpIter that was previously attached.
+func (ubr *unbufferedRegistration) detachCatchUpIter() *CatchUpIterator {
+	ubr.mu.Lock()
+	defer ubr.mu.Unlock()
+	catchUpIter := ubr.mu.catchUpIter
+	ubr.mu.catchUpIter = nil
+	return catchUpIter
+}
+
+// maybeRunCatchUpScan runs the catch-up scan if catchUpIter is not nil. It
+// promises to close catchUpIter once detached. It returns an error if catch-up
+// scan fails. Note that catch up scan bypasses BufferedStream and are sent to
+// underlying stream directly.
+func (ubr *unbufferedRegistration) maybeRunCatchUpScan(ctx context.Context) error {
+	catchUpIter := ubr.detachCatchUpIter()
+	if catchUpIter == nil {
+		return nil
+	}
+	start := timeutil.Now()
+	defer func() {
+		catchUpIter.Close()
+		ubr.metrics.RangeFeedCatchUpScanNanos.Inc(timeutil.Since(start).Nanoseconds())
+	}()
+
+	// In the future, we might want a separate queue for catch up scans.
```

Comment:
Interesting, is there any reason we couldn't send the catchup scan to the buffer other than wanting to avoid overflowing it with catchup events. Might be worth filing an issue for.

Reply:
No, sending catch up scan updates to the buffer is fine correctness-wise. We made a choice here to make catch up scan bypass buffer here https://cockroachlabs.slack.com/archives/C05JB0QURND/p1724282080131549?thread_ts=1724072786.336579&cid=C05JB0QURND. It is also the only reason why we need to have send unbuffered on buffered stream.